Check if the release exists before type casting

When a release doesn't exists causes a runtime error. Ensure the release exists before type casting

```
I0720 20:43:04.366106       1 util.go:107] Syncing cainelli-bunny-acme/nodejs-app-v1
E0720 20:43:04.366288       1 runtime.go:66] Observed a panic: &runtime.TypeAssertionError{interfaceString:"interface {}", concreteString:"", assertedString:"*v1alpha1.Release", missingMethod:""} (interface conversion: interface {} is nil, not *v1alpha1.Release)
/Users/san/work/GO/src/kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/Users/san/work/GO/src/kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/Users/san/work/GO/src/kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/iface.go:172
/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/build.go:114
/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/build.go:47
/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/util.go:108
/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/util.go:93
/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/util.go:73
/Users/san/work/GO/src/kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:96
/Users/san/work/GO/src/kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97
/Users/san/work/GO/src/kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52
/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/util.go:73
/usr/local/go/src/runtime/asm_amd64.s:2197
panic: interface conversion: interface {} is nil, not *v1alpha1.Release [recovered]
	panic: interface conversion: interface {} is nil, not *v1alpha1.Release

goroutine 120 [running]:
kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/Users/san/work/GO/src/kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x126
panic(0x12a60c0, 0xc42020f100)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
kolihub.io/koli/pkg/controller.(*BuildController).syncHandler(0xc4204e6190, 0xc4200428a0, 0x21, 0x1, 0x1)
	/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/build.go:114 +0xe5c
kolihub.io/koli/pkg/controller.(*BuildController).(kolihub.io/koli/pkg/controller.syncHandler)-fm(0xc4200428a0, 0x21, 0xa, 0xc42074ede8)
	/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/build.go:47 +0x3e
kolihub.io/koli/pkg/controller.(*TaskQueue).processNextWorkItem(0xc4201696e0)
	/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/util.go:108 +0x10f
kolihub.io/koli/pkg/controller.(*TaskQueue).runWorker(0xc4201696e0)
	/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/util.go:93 +0x2b
kolihub.io/koli/pkg/controller.(*TaskQueue).(kolihub.io/koli/pkg/controller.runWorker)-fm()
	/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/util.go:73 +0x2a
kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc420023fa8)
	/Users/san/work/GO/src/kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:96 +0x5e
kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc42075bfa8, 0x3b9aca00, 0x0, 0x7f992c178901, 0xc420082fc0)
	/Users/san/work/GO/src/kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97 +0xad
kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc420023fa8, 0x3b9aca00, 0xc420082fc0)
	/Users/san/work/GO/src/kolihub.io/koli/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52 +0x4d
kolihub.io/koli/pkg/controller.(*TaskQueue).run(0xc4201696e0, 0x3b9aca00, 0xc420082fc0)
	/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/util.go:73 +0x5e
created by kolihub.io/koli/pkg/controller.(*BuildController).Run
	/Users/san/work/GO/src/kolihub.io/koli/pkg/controller/build.go:100 +0x1f3
```

--- a/cmd/controller/controller-manager.go
+++ b/cmd/controller/controller-manager.go
@@ -41,6 +41,7 @@ func init() {
 	pflag.StringVar(&cfg.TLSConfig.CAFile, "ca-file", "", "path to TLS CA file.")
 	pflag.StringVar(&cfg.GitReleaseHost, "git-release-host", "http://git-release-server.koli-system", "the address where releases are stored")
 	pflag.StringVar(&cfg.ClusterName, "cluster-name", "gaia", "the name of the cluster")
+	pflag.StringVar(&cfg.PlatformJWTSecret, "platform-secret", "", "the jwt secret for creating dynamic system tokens")
 	pflag.StringVar(&cfg.SlugBuildImage, "slugbuilder-image", "quay.io/koli/slugbuilder", "the name of the builder image")
 	pflag.StringVar(&cfg.SlugRunnerImage, "slugrunner-image", "quay.io/koli/slugrunner", "the name of the runner image")
 	pflag.BoolVar(&cfg.DebugBuild, "debug-build", false, "debug the build container")
@@ -58,7 +59,10 @@ func printSystemVersion(kubeVersion *version.Info) {
 		v.GitVersion, v.GitCommit, v.GoVersion, v.BuildDate)
 }
 
-func startControllers(stop <-chan struct{}) error {
+func startControllers() error {
+	if len(cfg.PlatformJWTSecret) == 0 {
+		return fmt.Errorf("platform secret is empty")
+	}
 	kcfg, err := clientset.NewClusterConfig(cfg.Host, cfg.TLSInsecure, &cfg.TLSConfig)
 	if err != nil {
 		return err
@@ -87,33 +91,34 @@ func startControllers(stop <-chan struct{}) error {
 
 	sharedInformers := informers.NewSharedInformerFactory(client, 30*time.Second)
 
+	stopC := wait.NeverStop
 	// TODO: should we use the same client instance??
 	go controller.NewNamespaceController(
 		sharedInformers.Namespaces().Informer(),
 		sharedInformers.ServicePlans().Informer(sysClient),
 		client,
 		sysClient,
-	).Run(1, wait.NeverStop)
+	).Run(1, stopC)
 
 	go controller.NewAppManagerController(
 		sharedInformers.Deployments().Informer(),
 		sharedInformers.ServicePlans().Informer(sysClient),
 		client,
-	).Run(1, wait.NeverStop)
+	).Run(1, stopC)
 
 	go controller.NewReleaseController(
 		sharedInformers.Releases().Informer(sysClient),
 		sharedInformers.Deployments().Informer(),
 		sysClient,
 		client,
-	).Run(1, wait.NeverStop)
+	).Run(1, stopC)
 
 	go controller.NewBuildController(
 		&cfg,
 		sharedInformers.Releases().Informer(sysClient),
 		sysClient,
 		client,
-	).Run(1, wait.NeverStop)
+	).Run(1, stopC)
 
 	// TODO: hard-coded
 	selector := labels.Set{"kolihub.io/type": "slugbuild"}
@@ -124,10 +129,16 @@ func startControllers(stop <-chan struct{}) error {
 		sharedInformers.Releases().Informer(sysClient),
 		sysClient,
 		client,
-	).Run(1, wait.NeverStop)
+	).Run(1, stopC)
 
-	sharedInformers.Start(stop)
+	go controller.NewSecretController(
+		sharedInformers.Namespaces().Informer(),
+		sharedInformers.Secrets().Informer(),
+		client,
+		cfg.PlatformJWTSecret,
+	).Run(1, stopC)
 
+	sharedInformers.Start(stopC)
 	select {} // block forever
 }
 

--- a/cmd/controller/controller-manager.go
+++ b/cmd/controller/controller-manager.go
@@ -152,7 +152,8 @@ func main() {
 		}
 		fmt.Println(string(b))
 	} else {
-		err := startControllers(make(chan struct{}))
+		err := startControllers()
 		glog.Fatalf("error running controllers: %v", err)
+		panic("unreachable")
 	}
 }

--- a/cmd/controller/controller-manager.go
+++ b/cmd/controller/controller-manager.go
@@ -10,11 +10,11 @@ import (
 	"github.com/golang/glog"
 	"github.com/spf13/pflag"
 
+	_ "kolihub.io/koli/pkg/apis/v1alpha1/install"
 	"kolihub.io/koli/pkg/clientset"
 	"kolihub.io/koli/pkg/controller"
 	"kolihub.io/koli/pkg/controller/informers"
 	_ "kolihub.io/koli/pkg/controller/install"
-	_ "kolihub.io/koli/pkg/spec/install"
 	koliversion "kolihub.io/koli/pkg/version"
 
 	"k8s.io/apimachinery/pkg/labels"

--- a/pkg/apis/v1alpha1/constants.go
+++ b/pkg/apis/v1alpha1/constants.go
@@ -1,0 +1,7 @@
+package v1alpha1
+
+const (
+	// SystemSecretName it's a secret resource created and updated dynamically by a controller.
+	// It should be used for communicating between systems
+	SystemSecretName = "koli-system-token"
+)

--- a/pkg/apis/v1alpha1/meta.go
+++ b/pkg/apis/v1alpha1/meta.go
@@ -122,11 +122,7 @@ func (r *Release) GitCloneURL() (string, error) {
 
 // GitReleaseURL constructs the URL where the release must be stored
 func (r *Release) GitReleaseURL(host string) string {
-	repository := r.Spec.GitRepository
-	if r.IsGitHubSource() {
-		repository = filepath.Join(r.Namespace, r.Spec.DeployName)
-	}
-	urlPath := filepath.Join("releases", repository, r.Spec.GitRevision)
+	urlPath := filepath.Join("releases", r.Namespace, r.Spec.DeployName, r.Spec.GitRevision)
 	return fmt.Sprintf("%s/%s", host, urlPath)
 }
 

--- a/pkg/apis/v1alpha1/types.go
+++ b/pkg/apis/v1alpha1/types.go
@@ -137,16 +137,18 @@ const (
 // ReleaseSpec holds specification parameters of a release
 type ReleaseSpec struct {
 	// The URL of the git remote server to download the git revision tarball
-	GitRemote     string     `json:"gitRemote"`
-	GitRevision   string     `json:"gitRevision"`
-	GitRepository string     `json:"gitRepository"`
-	BuildRevision string     `json:"buildRevision"`
-	AutoDeploy    bool       `json:"autoDeploy"`
-	ExpireAfter   int32      `json:"expireAfter"`
-	DeployName    string     `json:"deployName"`
-	Build         bool       `json:"build"`
-	AuthToken     string     `json:"authToken"` // expirable token
-	Source        SourceType `json:"sourceType"`
+	GitRemote     string `json:"gitRemote"`
+	GitRevision   string `json:"gitRevision"`
+	GitRepository string `json:"gitRepository"`
+	BuildRevision string `json:"buildRevision"`
+	AutoDeploy    bool   `json:"autoDeploy"`
+	ExpireAfter   int32  `json:"expireAfter"`
+	DeployName    string `json:"deployName"`
+	Build         bool   `json:"build"`
+	// DEPRECATED, the authToken for each release is populated by a secret
+	// the lifecycle of the token is managed by a controller
+	AuthToken string     `json:"authToken"` // expirable token
+	Source    SourceType `json:"sourceType"`
 }
 
 // ReleaseList is a list of Release

--- a/pkg/apis/v1alpha1/well_known_annotations.go
+++ b/pkg/apis/v1alpha1/well_known_annotations.go
@@ -19,6 +19,8 @@ const (
 	AnnotationGitRevision = "kolihub.io/gitrevision"
 	// AnnotationGitBranch is the name of the branch to accept webhook requests
 	AnnotationGitBranch = "kolihub.io/gitbranch"
+	// DEPRECATED, the authToken for each release is populated by a secret
+	// the lifecycle of the token is managed by a controller
 	// AnnotationAuthToken it's a string credential to communication with the release server
 	AnnotationAuthToken = "kolihub.io/authtoken"
 	// AnnotationBuildSource it's the source of the request which triggered the build: github (webhook), local, gitstep, etc

--- a/pkg/apis/v1alpha1/well_known_annotations.go
+++ b/pkg/apis/v1alpha1/well_known_annotations.go
@@ -37,4 +37,7 @@ const (
 	AnnotationSetupStorage = "kolihub.io/setup-storage"
 	// AnnotationApp is used to identify the deployment from a parent resource
 	AnnotationApp = "kolihub.io/app"
+
+	// AnnotationSecretLastUpdated informs when a system token was last updated at
+	AnnotationSecretLastUpdated = "secret.kolihub.io/updated-at"
 )

--- a/pkg/apis/v1alpha1/well_known_labels.go
+++ b/pkg/apis/v1alpha1/well_known_labels.go
@@ -14,4 +14,7 @@ const (
 	LabelStoragePlan = "kolihub.io/storage-plan"
 	// LabelDefault indicates a resource as default (boolean)
 	LabelDefault = "kolihub.io/default"
+
+	// LabelSecretController identifies resources created by the secret controller
+	LabelSecretController = "secret.kolihub.io"
 )

--- a/pkg/controller/appmanager_test.go
+++ b/pkg/controller/appmanager_test.go
@@ -30,18 +30,21 @@ type fixture struct {
 	// Objects to put in the store.
 
 	objects []runtime.Object
+	secrets []runtime.Object
 	plans   []runtime.Object
 }
 
-func newFixture(t *testing.T, objects, plans []runtime.Object) *fixture {
+func newFixture(t *testing.T, objects, plans, secrets []runtime.Object) *fixture {
 	f := &fixture{}
 	f.t = t
 	f.objects = objects
 	f.plans = plans
+	f.secrets = secrets
+
 	return f
 }
 
-func (f *fixture) newController() (*AppManagerController, informers.SharedInformerFactory) {
+func (f *fixture) newAppController() (*AppManagerController, informers.SharedInformerFactory) {
 	f.client = fake.NewSimpleClientset(f.objects...)
 	informers := informers.NewSharedInformerFactory(f.client, 0)
 	ctrl := NewAppManagerController(
@@ -59,6 +62,24 @@ func (f *fixture) newController() (*AppManagerController, informers.SharedInform
 	return ctrl, informers
 }
 
+func (f *fixture) newSecretController(jwtToken string) (*SecretController, informers.SharedInformerFactory) {
+	f.client = fake.NewSimpleClientset(f.objects...)
+	infs := informers.NewSharedInformerFactory(f.client, 0)
+	ctrl := NewSecretController(
+		infs.Namespaces().Informer(),
+		infs.Secrets().Informer(),
+		f.client,
+		jwtToken,
+	)
+	for _, o := range f.objects {
+		ctrl.nsInf.GetStore().Add(o)
+	}
+	for _, o := range f.secrets {
+		ctrl.skInf.GetStore().Add(o)
+	}
+	return ctrl, infs
+}
+
 func TestSyncPersistentVolume(t *testing.T) {
 	var (
 		expectedName   = "foo"
@@ -68,8 +89,8 @@ func TestSyncPersistentVolume(t *testing.T) {
 		d              = newDeployment(expectedName, "dev-coyote-acme", notes, labels, nil, v1.Container{})
 		plan           = newStoragePlan("foo-plan-5g", "dev-coyote-acme", resource.MustParse("5Gi"))
 	)
-	f := newFixture(t, []runtime.Object{d}, []runtime.Object{plan})
-	c, _ := f.newController()
+	f := newFixture(t, []runtime.Object{d}, []runtime.Object{plan}, nil)
+	c, _ := f.newAppController()
 
 	if err := c.syncHandler(getKey(d, t)); err != nil {
 		t.Fatalf("unexpected error syncing: %v", err)
@@ -104,8 +125,8 @@ func TestSyncPersistentVolume(t *testing.T) {
 
 func TestSyncWithInvalidNamespace(t *testing.T) {
 	d := newDeployment("foo", "invalid-namespace", nil, nil, nil, v1.Container{})
-	f := newFixture(t, []runtime.Object{d}, nil)
-	c, _ := f.newController()
+	f := newFixture(t, []runtime.Object{d}, nil, nil)
+	c, _ := f.newAppController()
 
 	if err := c.syncHandler(getKey(d, t)); err != nil {
 		t.Fatalf("unexpected error syncing: %v", err)
@@ -121,8 +142,8 @@ func TestSyncWithNonExistentPlan(t *testing.T) {
 		d            = newDeployment("foo", "dev-coyote-acme", notes, labels, nil, v1.Container{})
 		plan         = newStoragePlan("wrong-storage-plan", "dev-coyote-acme", resource.MustParse("5Gi"))
 	)
-	f := newFixture(t, []runtime.Object{d}, []runtime.Object{plan})
-	c, _ := f.newController()
+	f := newFixture(t, []runtime.Object{d}, []runtime.Object{plan}, nil)
+	c, _ := f.newAppController()
 
 	if err := c.syncHandler(getKey(d, t)); err != nil && err.Error() != expectedMsg {
 		t.Errorf("GOT: %v, EXPECTED: %v", err, expectedMsg)

--- a/pkg/controller/build.go
+++ b/pkg/controller/build.go
@@ -16,6 +16,7 @@ import (
 	"k8s.io/client-go/tools/record"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 )
 
@@ -107,33 +108,38 @@ func (b *BuildController) Run(workers int, stopc <-chan struct{}) {
 }
 
 func (b *BuildController) syncHandler(key string) error {
-	obj, _, err := b.releaseInf.GetStore().GetByKey(key)
+	obj, exists, err := b.releaseInf.GetStore().GetByKey(key)
 	if err != nil {
 		return err
 	}
+	if err != nil {
+		return err
+	}
+	if !exists {
+		glog.V(3).Infof("%s - release doesn't exists, skip ...", key)
+		return nil
+	}
 	release := obj.(*platform.Release)
-	logHeader := fmt.Sprintf("%s/%s", release.Namespace, release.Name)
-
 	if release.DeletionTimestamp != nil {
 		// TODO: delete from the remote object store (minio/s3/gcs...)
-		glog.V(3).Infof("%s - release marked for deletion, skipping ...", logHeader)
+		glog.V(3).Infof("%s - release marked for deletion, skipping ...", key)
 		return nil
 	}
 	pns, err := platform.NewNamespace(release.Namespace)
 	if err != nil {
-		glog.V(3).Infof("%s - noop, it's not a valid namespace", logHeader)
+		glog.V(3).Infof("%s - noop, it's not a valid namespace", key)
 		return nil
 	}
 
 	if !release.Spec.Build {
-		glog.V(3).Infof("%s - noop, isn't a build action", logHeader)
+		glog.V(3).Infof("%s - noop, isn't a build action", key)
 		return nil
 	}
 
 	gitSha, err := koliutil.NewSha(release.Spec.GitRevision)
 	if err != nil {
 		// TODO: add an event informing the problem!
-		return fmt.Errorf("%s - %s", logHeader, err)
+		return fmt.Errorf("%s - %s", key, err)
 	}
 
 	info := koliutil.NewSlugBuilderInfo(
@@ -144,30 +150,20 @@ func (b *BuildController) syncHandler(key string) error {
 	sbPodName := fmt.Sprintf("sb-%s", release.Name)
 	pod, err := slugbuilderPod(b.config, release, gitSha, info)
 	if err != nil {
-		return fmt.Errorf("%s - failed creating slugbuild pod: %s", logHeader, err)
+		return fmt.Errorf("%s - failed creating slugbuild pod: %s", key, err)
 	}
 	_, err = b.kclient.Core().Pods(release.Namespace).Create(pod)
 	if err != nil {
 		// TODO: add an event informing the problem!
 		// TODO: requeue with backoff, got an error
-		return fmt.Errorf("%s - failed creating the slubuild pod: %s", logHeader, err)
+		return fmt.Errorf("%s - failed creating the slubuild pod: %s", key, err)
 	}
 
-	glog.Infof("%s - build started for '%s'", logHeader, sbPodName)
-	releaseCopy, err := platform.ReleaseDeepCopy(release)
-	if err != nil {
-		return fmt.Errorf("%s - failed deep copying release: %s", logHeader, err)
-	}
+	glog.Infof("%s - build started for '%s'", key, sbPodName)
 	// a build has started for this release, disable it!
-	releaseCopy.Spec.Build = false
-	// releaseCopy.TypeMeta = unversioned.TypeMeta{
-	// 	Kind:       "Release",
-	// 	APIVersion: spec.SchemeGroupVersion.String(),
-	// }
-
-	_, err = b.clientset.Release(releaseCopy.Namespace).Update(releaseCopy)
+	_, err = b.clientset.Release(release.Namespace).Patch(release.Name, types.MergePatchType, []byte(`{"spec": {"build": false}}`))
 	if err != nil {
-		return fmt.Errorf("%s - failed updating release: %s", logHeader, err)
+		return fmt.Errorf("%s - failed updating release: %s", key, err)
 	}
 	return nil
 }

--- a/pkg/controller/build.go
+++ b/pkg/controller/build.go
@@ -163,7 +163,7 @@ func (b *BuildController) syncHandler(key string) error {
 	// a build has started for this release, disable it!
 	_, err = b.clientset.Release(release.Namespace).Patch(release.Name, types.MergePatchType, []byte(`{"spec": {"build": false}}`))
 	if err != nil {
-		return fmt.Errorf("%s - failed updating release: %s", key, err)
+		return fmt.Errorf("failed updating release [%s]", err)
 	}
 	return nil
 }
@@ -181,7 +181,7 @@ func slugbuilderPod(cfg *Config, rel *platform.Release, gitSha *koliutil.SHA, in
 		"GIT_CLONE_URL":   gitCloneURL,
 		"GIT_RELEASE_URL": rel.GitReleaseURL(cfg.GitReleaseHost),
 		"GIT_REVISION":    rel.Spec.GitRevision,
-		"AUTH_TOKEN":      rel.Spec.AuthToken,
+		// "AUTH_TOKEN":      rel.Spec.AuthToken,
 	}
 	if cfg.DebugBuild {
 		env["DEBUG"] = "TRUE"
@@ -192,6 +192,18 @@ func slugbuilderPod(cfg *Config, rel *platform.Release, gitSha *koliutil.SHA, in
 	pod.Spec.Containers[0].Image = cfg.SlugBuildImage
 	pod.Spec.Containers[0].Name = rel.Name
 
+	// A dynamic JWT Token secret provisioned by a controller
+	pod.Spec.Containers[0].Env = append(pod.Spec.Containers[0].Env, v1.EnvVar{
+		Name: "AUTH_TOKEN",
+		ValueFrom: &v1.EnvVarSource{
+			SecretKeyRef: &v1.SecretKeySelector{
+				LocalObjectReference: v1.LocalObjectReference{
+					Name: platform.SystemSecretName,
+				},
+				Key: "token.jwt", // TODO: hard-coded
+			},
+		},
+	})
 	// addEnvToContainer(pod,a "PUT_PATH", info.PushKey(), 0)
 	return &pod, nil
 }
@@ -229,14 +241,12 @@ func podResource(rel *platform.Release, gitSha *koliutil.SHA, env map[string]int
 		}
 	}
 
-	if len(pod.Spec.Containers) > 0 {
-		for k, v := range env {
-			for i := range pod.Spec.Containers {
-				pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, v1.EnvVar{
-					Name:  k,
-					Value: fmt.Sprintf("%v", v),
-				})
-			}
+	for k, v := range env {
+		for i := range pod.Spec.Containers {
+			pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, v1.EnvVar{
+				Name:  k,
+				Value: fmt.Sprintf("%v", v),
+			})
 		}
 	}
 	return pod

--- a/pkg/controller/config.go
+++ b/pkg/controller/config.go
@@ -4,14 +4,15 @@ import "k8s.io/client-go/rest"
 
 // Config defines configuration parameters for the Operator.
 type Config struct {
-	Host            string
-	GitReleaseHost  string
-	TLSInsecure     bool
-	TLSConfig       rest.TLSClientConfig
-	SlugBuildImage  string
-	SlugRunnerImage string
-	ClusterName     string
-	DebugBuild      bool
+	Host              string
+	GitReleaseHost    string
+	TLSInsecure       bool
+	TLSConfig         rest.TLSClientConfig
+	SlugBuildImage    string
+	SlugRunnerImage   string
+	PlatformJWTSecret string
+	ClusterName       string
+	DebugBuild        bool
 }
 
 // IsValidStorageType check if it's valid storage type

--- a/pkg/controller/deployer.go
+++ b/pkg/controller/deployer.go
@@ -15,6 +15,7 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/tools/record"
 
+	"k8s.io/apimachinery/pkg/types"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/client-go/pkg/api/v1"
 	extensions "k8s.io/client-go/pkg/apis/extensions/v1beta1"
@@ -197,15 +198,11 @@ func (d *DeployerController) syncHandler(key string) error {
 			"Deployed", "Deploy '%s' updated with the new revision [%s]", release.Spec.DeployName, release.Spec.GitRevision[:7])
 	}
 
-	releaseCopy, err := platform.ReleaseDeepCopy(release)
-	if err != nil {
-		return fmt.Errorf("failed deep copying [%s]", err)
-	}
 	// turn-off build, otherwise it will trigger unwanted builds
-	releaseCopy.Spec.Build = false
-	releaseCopy.Labels[spec.KoliPrefix("buildstatus")] = string(pod.Status.Phase)
-	if _, err := d.clientset.Release(pod.Namespace).Update(releaseCopy); err != nil {
-		return fmt.Errorf("failed updating pod status [%s]", err)
+	patchData := []byte(fmt.Sprintf(`{"metadata": {"labels": {"kolihub.io/buildstatus": "%v"}}, "spec": {"build": false}}`, pod.Status.Phase))
+	_, err = d.clientset.Release(pod.Namespace).Patch(release.Name, types.MergePatchType, patchData)
+	if err != nil {
+		return fmt.Errorf("failed patching pod status[%s] in release [%s]", pod.Status.Phase, err)
 	}
 	return nil
 }
@@ -238,9 +235,17 @@ func (d *DeployerController) deploySlug(release *platform.Release, deploy *exten
 			Name:  "SLUG_URL",
 			Value: slugURL,
 		},
+		// A dynamic JWT Token secret provisioned by a controller
 		{
-			Name:  "AUTH_TOKEN",
-			Value: release.Spec.AuthToken, // TODO: this token mustn't be expirable!
+			Name: "AUTH_TOKEN",
+			ValueFrom: &v1.EnvVarSource{
+				SecretKeyRef: &v1.SecretKeySelector{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: platform.SystemSecretName,
+					},
+					Key: "token.jwt", // TODO: hard-coded
+				},
+			},
 		},
 		{
 			Name:  "DEBUG",

--- a/pkg/controller/informers/core.go
+++ b/pkg/controller/informers/core.go
@@ -274,11 +274,11 @@ func (f *secretInformer) Informer() cache.SharedIndexInformer {
 		&cache.ListWatch{
 			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
 				options.LabelSelector = fmt.Sprintf("%s=true", platform.LabelSecretController)
-				return f.client.Core().Pods(metav1.NamespaceAll).List(options)
+				return f.client.Core().Secrets(metav1.NamespaceAll).List(options)
 			},
 			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
 				options.LabelSelector = fmt.Sprintf("%s=true", platform.LabelSecretController)
-				return f.client.Core().Pods(metav1.NamespaceAll).Watch(options)
+				return f.client.Core().Secrets(metav1.NamespaceAll).Watch(options)
 			},
 		},
 		&v1.Secret{},

--- a/pkg/controller/informers/core.go
+++ b/pkg/controller/informers/core.go
@@ -1,6 +1,7 @@
 package informers
 
 import (
+	"fmt"
 	"reflect"
 
 	platform "kolihub.io/koli/pkg/apis/v1alpha1"
@@ -245,6 +246,44 @@ func (f *podInformer) Informer(selector labels.Selector) cache.SharedIndexInform
 		&v1.Pod{},
 		f.defaultResync,
 		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+	)
+	f.informers[informerType] = informer
+	return informer
+}
+
+// SecretInformer is a type of SharedIndexInformer which watches and lists all its resources
+type SecretInformer interface {
+	Informer() cache.SharedIndexInformer
+}
+
+type secretInformer struct {
+	*sharedInformerFactory
+}
+
+func (f *secretInformer) Informer() cache.SharedIndexInformer {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	informerType := reflect.TypeOf(&v1.Secret{})
+	informer, exists := f.informers[informerType]
+	if exists {
+		return informer
+	}
+
+	informer = cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.LabelSelector = fmt.Sprintf("%s=true", platform.LabelSecretController)
+				return f.client.Core().Pods(metav1.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = fmt.Sprintf("%s=true", platform.LabelSecretController)
+				return f.client.Core().Pods(metav1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Secret{},
+		f.defaultResync,
+		cache.Indexers{},
 	)
 	f.informers[informerType] = informer
 	return informer

--- a/pkg/controller/informers/informer.go
+++ b/pkg/controller/informers/informer.go
@@ -23,6 +23,7 @@ type SharedInformerFactory interface {
 	Namespaces() NamespaceInformer
 	Deployments() DeploymentInformer
 	Pods() PodInformer
+	Secrets() SecretInformer
 }
 
 type sharedInformerFactory struct {
@@ -119,4 +120,9 @@ func (f *sharedInformerFactory) Namespaces() NamespaceInformer {
 // Pods returns a SharedIndexInformer that lists and watchs all pods
 func (f *sharedInformerFactory) Pods() PodInformer {
 	return &podInformer{sharedInformerFactory: f}
+}
+
+// Secrets returns a SharedIndexInformer that lists and watchs all secrets
+func (f *sharedInformerFactory) Secrets() SecretInformer {
+	return &secretInformer{sharedInformerFactory: f}
 }

--- a/pkg/controller/release.go
+++ b/pkg/controller/release.go
@@ -122,6 +122,7 @@ func (r *ReleaseController) syncHandler(key string) error {
 	}
 
 	dp := obj.(*extensions.Deployment)
+	// TODO: change to namespace meta!
 	_, err = platform.NewNamespace(dp.Namespace)
 	if err != nil {
 		glog.V(2).Infof("%s - noop, it's not a valid namespace", key)
@@ -247,6 +248,6 @@ func activateBuildPayload(activate bool) []byte {
 	if activate {
 		build = "true"
 	}
-	payload := fmt.Sprintf(`{"metadata": {"annotations": {"%s": "%s"}}}`, spec.KoliPrefix("build"), build)
+	payload := fmt.Sprintf(`{"metadata": {"annotations": {"%s": "%s"}}}`, platform.AnnotationBuild, build)
 	return []byte(payload)
 }

--- a/pkg/controller/secrets.go
+++ b/pkg/controller/secrets.go
@@ -142,7 +142,7 @@ func (c *SecretController) syncHandler(key string) error {
 
 	// Patching an object primarily is less expensive because the most part of
 	// secret resources will be synced in each resync
-	s, err := c.kclient.Core().Secrets(ns.Name).Patch(
+	_, err = c.kclient.Core().Secrets(ns.Name).Patch(
 		platform.SystemSecretName,
 		types.StrategicMergePatchType,
 		genSystemTokenPatchData(tokenString),
@@ -150,8 +150,8 @@ func (c *SecretController) syncHandler(key string) error {
 	if err != nil && !apierrors.IsNotFound(err) {
 		return fmt.Errorf("failed updating secret [%v]", err)
 	}
-	if s != nil {
-		glog.V(3).Infof(`%s/%s secret updated with success, `, key, s.Name)
+	if err == nil {
+		glog.V(3).Infof(`%s/%s secret updated with success`, key, platform.SystemSecretName)
 		return nil
 	}
 

--- a/pkg/controller/secrets.go
+++ b/pkg/controller/secrets.go
@@ -114,7 +114,12 @@ func (c *SecretController) syncHandler(key string) error {
 	// Verify the last time an object was synced, using a shared informer doesn't permit having a
 	// custom resync time, thus veryfing if an object can perform a sync operation is important.
 	// This validation prevents starving resources from the api server
-	if obj, exists, _ = c.skInf.GetStore().GetByKey(fmt.Sprintf("%s/%s", key, platform.SystemSecretName)); exists {
+	obj, exists, err = c.skInf.GetStore().GetByKey(fmt.Sprintf("%s/%s", key, platform.SystemSecretName))
+	if err != nil {
+		glog.Warningf("%s/%s - failed retrieving secret from store [%v]", key, platform.SystemSecretName, err)
+	}
+	if exists {
+		glog.V(5).Infof("%s/%s - secret exists in cache", key, platform.SystemSecretName)
 		s := obj.(*v1.Secret)
 		if s.Annotations != nil {
 			lastUpdated, err := time.Parse(time.RFC3339, s.Annotations[platform.AnnotationSecretLastUpdated])
@@ -152,7 +157,7 @@ func (c *SecretController) syncHandler(key string) error {
 		return fmt.Errorf("failed updating secret [%v]", err)
 	}
 	if err == nil {
-		glog.V(3).Infof(`%s/%s secret updated with success`, key, platform.SystemSecretName)
+		glog.Infof(`%s/%s secret updated with success`, key, platform.SystemSecretName)
 		return nil
 	}
 
@@ -173,7 +178,7 @@ func (c *SecretController) syncHandler(key string) error {
 	if err != nil {
 		return fmt.Errorf("failed creating secret [%v]", err)
 	}
-	glog.V(3).Infof("%s/%s secret created with success", key, platform.SystemSecretName)
+	glog.Infof("%s/%s secret created with success", key, platform.SystemSecretName)
 	return nil
 }
 

--- a/pkg/controller/secrets.go
+++ b/pkg/controller/secrets.go
@@ -2,6 +2,7 @@ package controller
 
 import (
 	"bytes"
+	"encoding/base64"
 	"fmt"
 	"time"
 
@@ -183,7 +184,7 @@ func genSystemTokenPatchData(token string) []byte {
 			platform.LabelSecretController,
 			platform.AnnotationSecretLastUpdated,
 			time.Now().UTC().Format(time.RFC3339),
-			token,
+			base64.StdEncoding.EncodeToString(bytes.NewBufferString(token).Bytes()),
 		),
 	)
 }

--- a/pkg/controller/secrets.go
+++ b/pkg/controller/secrets.go
@@ -74,7 +74,7 @@ func (c *SecretController) Run(workers int, stopc <-chan struct{}) {
 
 	glog.Info("Starting Secret controller...")
 
-	if !cache.WaitForCacheSync(stopc, c.nsInf.HasSynced) {
+	if !cache.WaitForCacheSync(stopc, c.nsInf.HasSynced, c.skInf.HasSynced) {
 		return
 	}
 

--- a/pkg/controller/secrets.go
+++ b/pkg/controller/secrets.go
@@ -1,0 +1,189 @@
+package controller
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+
+	platform "kolihub.io/koli/pkg/apis/v1alpha1"
+	"kolihub.io/koli/pkg/apis/v1alpha1/draft"
+	"kolihub.io/koli/pkg/util"
+)
+
+// SecretController generates dynamic secrets through namespaces. It manages secrets
+// which have specific expiration time, usually used for machine-to-machine communication
+// with limited access throughout the platform
+type SecretController struct {
+	kclient kubernetes.Interface
+
+	nsInf cache.SharedIndexInformer
+	skInf cache.SharedIndexInformer
+
+	queue    *TaskQueue
+	recorder record.EventRecorder
+
+	jwtSecret string
+}
+
+// NewSecretController creates a new SecretController
+func NewSecretController(nsInf, skInf cache.SharedIndexInformer, client kubernetes.Interface, jwtSecret string) *SecretController {
+	c := &SecretController{
+		kclient:   client,
+		nsInf:     nsInf,
+		skInf:     skInf,
+		recorder:  newRecorder(client, "secret-controller"),
+		jwtSecret: jwtSecret,
+	}
+	c.queue = NewTaskQueue(c.syncHandler)
+	c.nsInf.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addNamespace,
+		UpdateFunc: c.updateNamespace,
+	})
+	return c
+}
+
+func (c *SecretController) addNamespace(n interface{}) {
+	new := n.(*v1.Namespace)
+	glog.V(2).Infof("%s(%s) - add-namespace", new.Name, new.ResourceVersion)
+	c.queue.Add(new)
+}
+
+func (c *SecretController) updateNamespace(o, n interface{}) {
+	new := n.(*v1.Namespace)
+	glog.V(2).Infof("%s(%s) - update-namespace", new.Name, new.ResourceVersion)
+	c.queue.Add(new)
+}
+
+// Run the controller.
+func (c *SecretController) Run(workers int, stopc <-chan struct{}) {
+	// don't let panics crash the process
+	defer utilruntime.HandleCrash()
+	// make sure the work queue is shutdown which will trigger workers to end
+	defer c.queue.shutdown()
+
+	glog.Info("Starting Secret controller...")
+
+	if !cache.WaitForCacheSync(stopc, c.nsInf.HasSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		// runWorker will loop until "something bad" happens.
+		// The .Until will then rekick the worker after one second
+		go c.queue.run(time.Second, stopc)
+	}
+
+	// wait until we're told to stop
+	<-stopc
+	glog.Info("Shutting down Secret controller")
+}
+
+func (c *SecretController) syncHandler(key string) error {
+	obj, exists, err := c.nsInf.GetStore().GetByKey(key)
+	if err != nil {
+		glog.Warningf("%s - failed retrieving object from store [%s]", key, err)
+		return err
+	}
+	if !exists {
+		glog.V(3).Infof("%s - the namespace doesn't exists", key)
+		return nil
+	}
+	ns := obj.(*v1.Namespace)
+	if ns.DeletionTimestamp != nil {
+		glog.V(3).Infof("%s - object marked for deletion, skip ...", key)
+		return nil
+	}
+
+	nsMeta := draft.NewNamespaceMetadata(ns.Name)
+	if !nsMeta.IsValid() {
+		glog.V(3).Infof("%s - it's not a valid namespace, skip ...", key)
+		return nil
+	}
+
+	// Verify the last time an object was synced, using a shared informer doesn't permit having a
+	// custom resync time, thus veryfing if an object can perform a sync operation is important.
+	// This validation prevents starving resources from the api server
+	if obj, exists, _ = c.skInf.GetStore().GetByKey(fmt.Sprintf("%s/%s", key, platform.SystemSecretName)); exists {
+		s := obj.(*v1.Secret)
+		if s.Annotations != nil {
+			lastUpdated, err := time.Parse(time.RFC3339, s.Annotations[platform.AnnotationSecretLastUpdated])
+			if err == nil && lastUpdated.Add(time.Minute*20).After(time.Now().UTC()) {
+				glog.V(3).Infof(`%s/%s - to soon for updating secret "%s"`, key, platform.SystemSecretName, lastUpdated.Format(time.RFC3339))
+				return nil
+			}
+			if err != nil {
+				glog.Warningf("%s - got error converting time [%v]", key, err)
+			}
+		}
+	}
+	// Generate a system token based on the customer and organization of the namespace.
+	// The access token has limited access to the resources of the platform
+	tokenString, err := util.GenerateNewJwtToken(
+		c.jwtSecret,
+		nsMeta.Customer(),
+		nsMeta.Organization(),
+		platform.SystemTokenType,
+		time.Now().UTC().Add(time.Hour*1), // hard-coded exp time
+	)
+	if err != nil {
+		glog.Warningf("%s - failed generating system token [%v]", key, err)
+		return nil
+	}
+
+	// Patching an object primarily is less expensive because the most part of
+	// secret resources will be synced in each resync
+	s, err := c.kclient.Core().Secrets(ns.Name).Patch(
+		platform.SystemSecretName,
+		types.StrategicMergePatchType,
+		genSystemTokenPatchData(tokenString),
+	)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("failed updating secret [%v]", err)
+	}
+	if s != nil {
+		glog.V(3).Infof(`%s/%s secret updated with success, `, key, s.Name)
+		return nil
+	}
+
+	_, err = c.kclient.Core().Secrets(ns.Name).Create(&v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns.Name,
+			Name:      platform.SystemSecretName,
+			Labels:    map[string]string{platform.LabelSecretController: "true"},
+			Annotations: map[string]string{
+				platform.AnnotationSecretLastUpdated: time.Now().UTC().Format(time.RFC3339),
+			},
+		},
+		Data: map[string][]byte{
+			"token.jwt": bytes.NewBufferString(tokenString).Bytes(),
+		},
+		Type: v1.SecretTypeOpaque,
+	})
+	if err != nil {
+		return fmt.Errorf("failed creating secret [%v]", err)
+	}
+	glog.V(3).Infof("%s/%s secret created with success", key, platform.SystemSecretName)
+	return nil
+}
+
+func genSystemTokenPatchData(token string) []byte {
+	return []byte(
+		fmt.Sprintf(
+			`{"metadata": {"labels": {"%s": "true"}, "annotations": {"%s": "%s"}}, "data": {"token.jwt": "%s"}, "type": "Opaque"}`,
+			platform.LabelSecretController,
+			platform.AnnotationSecretLastUpdated,
+			time.Now().UTC().Format(time.RFC3339),
+			token,
+		),
+	)
+}

--- a/pkg/controller/secrets_test.go
+++ b/pkg/controller/secrets_test.go
@@ -1,0 +1,149 @@
+package controller
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/pkg/api/v1"
+	core "k8s.io/client-go/testing"
+	platform "kolihub.io/koli/pkg/apis/v1alpha1"
+	"kolihub.io/koli/pkg/util"
+)
+
+func newSecret(name, ns, token string, lastUpdate time.Time) *v1.Secret {
+	return &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns,
+			Name:      name,
+			Labels:    map[string]string{platform.LabelSecretController: "true"},
+			Annotations: map[string]string{
+				platform.AnnotationSecretLastUpdated: lastUpdate.Format(time.RFC3339),
+			},
+		},
+		Data: map[string][]byte{
+			"token.jwt": bytes.NewBufferString(token).Bytes(),
+		},
+		Type: v1.SecretTypeOpaque,
+	}
+}
+
+func newNamespace(name string) *v1.Namespace {
+	return &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+	}
+}
+
+func TestSyncSecrets(t *testing.T) {
+	var (
+		customer, org, jwtSecret = "coyote", "acme", "asecret"
+		ns                       = fmt.Sprintf("dev-%s-%s", customer, org)
+		expectedObj              = newSecret(platform.SystemSecretName, ns, "", time.Now().UTC())
+	)
+	tokenStr, _ := util.GenerateNewJwtToken(jwtSecret, customer, org, platform.SystemTokenType, time.Now().UTC().Add(time.Hour*1))
+	expectedObj.Data["token.jwt"] = bytes.NewBufferString(tokenStr).Bytes()
+	f := newFixture(t, []runtime.Object{newNamespace(ns)}, nil, nil)
+	c, _ := f.newSecretController("asecret")
+	f.client.AddReactor("patch", "secrets", func(action core.Action) (bool, runtime.Object, error) {
+		return true, nil, apierrors.NewNotFound(action.GetResource().GroupResource(), action.GetResource().Resource)
+	})
+	if err := c.syncHandler(ns); err != nil {
+		t.Fatalf("got unexpected error: %v", err)
+	}
+
+	if len(f.client.Actions()) != 2 {
+		t.Fatalf("unexpected length of action(s), found %d action(s)", len(f.client.Actions()))
+	}
+
+	for _, action := range f.client.Actions() {
+		switch a := action.(type) {
+		case core.PatchActionImpl: // no-op, simulated a not found resource
+		case core.CreateActionImpl:
+			s := a.GetObject().(*v1.Secret)
+			if !reflect.DeepEqual(s, expectedObj) {
+				t.Errorf("GOT: %#v, EXPECTED: %#v", s, expectedObj)
+			}
+		default:
+			t.Errorf("unexpected type of action: %T, OBJ: %v", a, action)
+		}
+	}
+}
+
+func TestSyncSecretsWithValidLastUpdatedTime(t *testing.T) {
+	var (
+		customer, org = "coyote", "acme"
+		ns            = fmt.Sprintf("dev-%s-%s", customer, org)
+		expectedObj   = newSecret(platform.SystemSecretName, ns, "", time.Now().UTC())
+	)
+	f := newFixture(t, []runtime.Object{newNamespace(ns)}, nil, []runtime.Object{expectedObj})
+	c, _ := f.newSecretController("")
+	if err := c.syncHandler(ns); err != nil {
+		t.Fatalf("got unexpected error: %v", err)
+	}
+	if len(f.client.Actions()) != 0 {
+		t.Fatalf("unexpected length of action(s), got %d action(s), expected 0", len(f.client.Actions()))
+	}
+}
+
+func TestSyncSecretsWithExpiredLastUpdatedTime(t *testing.T) {
+	var (
+		customer, org = "coyote", "acme"
+		ns            = fmt.Sprintf("dev-%s-%s", customer, org)
+		expectedObj   = newSecret(platform.SystemSecretName, ns, "", time.Now().UTC().Add(time.Minute-time.Minute*25))
+	)
+	f := newFixture(t, []runtime.Object{newNamespace(ns)}, nil, []runtime.Object{expectedObj})
+	c, _ := f.newSecretController("")
+	if err := c.syncHandler(ns); err != nil {
+		t.Fatalf("got unexpected error: %v", err)
+	}
+	if len(f.client.Actions()) != 1 {
+		t.Fatalf("unexpected length of action(s), got %d action(s), expected 1", len(f.client.Actions()))
+	}
+}
+
+func TestSyncSecretsWithErrorOnPatch(t *testing.T) {
+	var (
+		customer, org = "coyote", "acme"
+		ns            = fmt.Sprintf("dev-%s-%s", customer, org)
+		expectedObj   = newSecret(platform.SystemSecretName, ns, "", time.Now().UTC().Add(time.Minute-time.Minute*25))
+		expError      = fmt.Errorf("failed updating secret [bad request happened]")
+	)
+	f := newFixture(t, []runtime.Object{newNamespace(ns)}, nil, []runtime.Object{expectedObj})
+	c, _ := f.newSecretController("")
+	f.client.AddReactor("patch", "secrets", func(action core.Action) (bool, runtime.Object, error) {
+		return true, nil, apierrors.NewBadRequest("bad request happened")
+	})
+	err := c.syncHandler(ns)
+	if !reflect.DeepEqual(err, expError) {
+		t.Errorf("GOT: %#v, EXPECTED: %#v", err, expError)
+	}
+}
+
+func TestSyncSecretsWithErrorOnCreate(t *testing.T) {
+	var (
+		customer, org = "coyote", "acme"
+		ns            = fmt.Sprintf("dev-%s-%s", customer, org)
+		expectedObj   = newSecret(platform.SystemSecretName, ns, "", time.Now().UTC().Add(time.Minute-time.Minute*25))
+		expError      = fmt.Errorf("failed creating secret [bad request happened]")
+	)
+	f := newFixture(t, []runtime.Object{newNamespace(ns)}, nil, []runtime.Object{expectedObj})
+	c, _ := f.newSecretController("")
+	f.client.AddReactor("patch", "secrets", func(action core.Action) (bool, runtime.Object, error) {
+		return true, nil, apierrors.NewNotFound(action.GetResource().GroupResource(), action.GetResource().Resource)
+	})
+
+	f.client.PrependReactor("create", "secrets", func(action core.Action) (bool, runtime.Object, error) {
+		return true, nil, apierrors.NewBadRequest("bad request happened")
+	})
+	err := c.syncHandler(ns)
+	if !reflect.DeepEqual(err, expError) {
+		t.Errorf("GOT: %#v, EXPECTED: %#v", err, expError)
+	}
+}

--- a/pkg/git/api/authenticate.go
+++ b/pkg/git/api/authenticate.go
@@ -31,9 +31,8 @@ func (h *Handler) Authenticate(w http.ResponseWriter, r *http.Request, next http
 	}
 	h.user = u
 
-	// A system token is only allowed to download releases,
+	// A system token is only allowed to download/upload releases,
 	if u.Type == platform.SystemTokenType &&
-		r.Method != "GET" &&
 		!releasesRegexp.MatchString(r.URL.Path) {
 		w.WriteHeader(http.StatusForbidden)
 		fmt.Fprintf(w, "Access Denied! Not allowed to access the resource\n")

--- a/pkg/git/api/github.go
+++ b/pkg/git/api/github.go
@@ -25,7 +25,6 @@ import (
 	"kolihub.io/koli/pkg/apis/v1alpha1/draft"
 	"kolihub.io/koli/pkg/clientset/auth0"
 	auth0clientset "kolihub.io/koli/pkg/clientset/auth0"
-	gitutil "kolihub.io/koli/pkg/git/util"
 	"kolihub.io/koli/pkg/util"
 )
 
@@ -501,6 +500,7 @@ func (h *Handler) Webhooks(w http.ResponseWriter, r *http.Request) {
 
 	switch event := event.(type) {
 	case *github.PushEvent:
+		glog.Infof("got a push event[%d], pusher[%#s], repository[%#v]", event.GetPushID(), event.Pusher.GetEmail(), event.Repo.GetFullName())
 		l := labels.Set{
 			"kolihub.io/gitowner": event.Repo.Owner.GetName(),
 			"kolihub.io/gitrepo":  event.Repo.GetName(),
@@ -512,7 +512,7 @@ func (h *Handler) Webhooks(w http.ResponseWriter, r *http.Request) {
 			fmt.Fprintf(w, "failed retrieving deploys: %s\n", err)
 			return
 		}
-
+		glog.Infof("got %d candidate(s) for build", len(dList.Items))
 		for _, obj := range dList.Items {
 			dp := draft.NewDeployment(&obj)
 			target := filepath.Join(dp.Namespace, dp.Name)
@@ -527,6 +527,7 @@ func (h *Handler) Webhooks(w http.ResponseWriter, r *http.Request) {
 				glog.Infof("%s - branch is not related: %s\n", target, dp.GitBranch())
 				continue
 			}
+			glog.Infof("%s - processing webhook", target)
 
 			// Fetch the identity only if the repository is private
 			var identity *authentication.Identity
@@ -544,11 +545,11 @@ func (h *Handler) Webhooks(w http.ResponseWriter, r *http.Request) {
 				}
 			}
 
-			jwtSystemToken, err := gitutil.GenerateNewJwtToken(h.cnf.PlatformClientSecret, nsMeta.Customer(), nsMeta.Organization(), platform.SystemTokenType)
-			if err != nil {
-				glog.Infof("%s - failed generating user token: %s", target, err)
-				continue
-			}
+			// jwtSystemToken, err := gitutil.GenerateNewJwtToken(h.cnf.PlatformClientSecret, nsMeta.Customer(), nsMeta.Organization(), platform.SystemTokenType)
+			// if err != nil {
+			// 	glog.Infof("%s - failed generating user token: %s", target, err)
+			// 	continue
+			// }
 
 			original, err := dp.DeepCopy()
 			if err != nil {
@@ -559,8 +560,8 @@ func (h *Handler) Webhooks(w http.ResponseWriter, r *http.Request) {
 
 			// Always remove at this stage, if the repository is public
 			// the token is not required for cloning.
-			delete(dp.Annotations, platform.AnnotationAuthToken)
-			delete(dp.Annotations, "kolihub.io/authtokentype")
+			// delete(dp.Annotations, platform.AnnotationAuthToken)
+			// delete(dp.Annotations, "kolihub.io/authtokentype")
 
 			// TODO: check if kolihub.io/build == true, means that a build was already started
 			// TODO: try to recover the revision number from the releases resources
@@ -571,7 +572,7 @@ func (h *Handler) Webhooks(w http.ResponseWriter, r *http.Request) {
 			dp.Annotations[platform.AnnotationGitRemote] = event.Repo.GetCloneURL()
 			dp.Annotations[platform.AnnotationGitRepository] = event.Repo.GetFullName()
 			dp.Annotations[platform.AnnotationGitRevision] = event.HeadCommit.GetID()
-			dp.Annotations[platform.AnnotationAuthToken] = jwtSystemToken
+			// dp.Annotations[platform.AnnotationAuthToken] = jwtSystemToken
 			if identity != nil {
 				cloneURL, err := getCloneURL(event.Repo.GetCloneURL(), identity.AccessToken, event.Repo.GetFullName())
 				if err != nil {

--- a/pkg/git/api/github_test.go
+++ b/pkg/git/api/github_test.go
@@ -419,7 +419,6 @@ func TestWebhookDeployOneApp(t *testing.T) {
 				platform.AnnotationGitRepository: path.Join(ownerRepo, repoName),
 				platform.AnnotationGitRevision:   commit,
 				platform.AnnotationBuildSource:   "github",
-				platform.AnnotationAuthToken:     "",
 				platform.AnnotationBuild:         "true",
 			},
 		}
@@ -455,7 +454,6 @@ func TestWebhookDeployOneApp(t *testing.T) {
 			if err := json.Unmarshal(tp.Patch, d); err != nil {
 				t.Fatalf("unexpected error unmarshaling deploy: %#v", err)
 			}
-			expectedDeployMeta.Annotations[platform.AnnotationAuthToken] = d.Annotations[platform.AnnotationAuthToken]
 			if !reflect.DeepEqual(expectedDeployMeta.Annotations, d.Annotations) {
 				t.Errorf("GOT: %#v, EXPECTED: %#v", d.Annotations, expectedDeployMeta.Annotations)
 			}

--- a/pkg/git/api/releases.go
+++ b/pkg/git/api/releases.go
@@ -10,6 +10,7 @@ import (
 	"path/filepath"
 	"strconv"
 
+	"github.com/golang/glog"
 	"github.com/gorilla/mux"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"kolihub.io/koli/pkg/apis/v1alpha1/draft"
@@ -21,8 +22,14 @@ import (
 func (h *Handler) Releases(w http.ResponseWriter, r *http.Request) {
 	params := mux.Vars(r)
 	namespace, deployName, gitRev := params["namespace"], params["deployName"], params["gitSha"]
+	key := fmt.Sprintf("%s/%s", namespace, deployName)
 	if _, err := draft.NewSha(gitRev); err != nil {
 		util.WriteResponseError(w, util.StatusBadRequest("invalid git SHA format", nil, metav1.StatusReasonBadRequest))
+		return
+	}
+	nsMeta := draft.NewNamespaceMetadata(namespace)
+	if nsMeta.Customer() != h.user.Customer || nsMeta.Organization() != h.user.Organization {
+		util.WriteResponseError(w, util.StatusForbidden("the user is not the owner of the namespace", nil, metav1.StatusReasonForbidden))
 		return
 	}
 	// path := filepath.Join(h.cnf.GitHome, constants.GitReleasePath, namespace, deployName)
@@ -55,12 +62,8 @@ func (h *Handler) Releases(w http.ResponseWriter, r *http.Request) {
 
 		filePath := filepath.Join(gitTask.FullReleasePath(), gitRev, header.Filename)
 		if _, err := os.Stat(filePath); !os.IsNotExist(err) {
-			util.WriteResponseError(w, &metav1.Status{
-				Code:    http.StatusConflict,
-				Status:  metav1.StatusFailure,
-				Message: fmt.Sprintf("the file %s already exists for this revision", header.Filename),
-				Reason:  metav1.StatusReasonConflict,
-			})
+			glog.Infof(`%s - the file "%s" already exists for this revision, noop`, key, header.Filename)
+			util.WriteResponseNoContent(w)
 			return
 		}
 		io.Copy(&data, file)

--- a/pkg/git/receive/gitreceive.go
+++ b/pkg/git/receive/gitreceive.go
@@ -56,12 +56,11 @@ func Run(conf *Config, oldRev, newRev, refName string) error {
 	// }
 	gitTask := gitutil.NewServerTask(conf.GitHome, gitutil.NewObjectMeta(dp.Name, dp.Namespace))
 
-	// TODO: use a system token instead of the user
 	dp.Annotations[platform.AnnotationBuild] = "true"
 	dp.Annotations[platform.AnnotationGitRepository] = gitTask.GetRepository()
 	dp.Annotations[platform.AnnotationGitRemote] = conf.GitAPIHostname
 	dp.Annotations[platform.AnnotationGitRevision] = gitSha.Full()
-	dp.Annotations[platform.AnnotationAuthToken] = conf.UserJwtToken
+	// dp.Annotations[platform.AnnotationAuthToken] = conf.UserJwtToken
 	dp.Annotations[platform.AnnotationBuildSource] = "local"
 
 	if _, err := client.Extensions().Deployments(dp.Namespace).Update(dp.GetObject()); err != nil {

--- a/pkg/git/server/authenticate.go
+++ b/pkg/git/server/authenticate.go
@@ -25,7 +25,7 @@ func (h *Handler) Authenticate(w http.ResponseWriter, r *http.Request, next http
 		return
 	}
 	h.user = u
-	// A system token is only allowed to download releases,
+	// A system token is only allowed to download/upload releases,
 	if u.Type == platform.SystemTokenType {
 		w.WriteHeader(http.StatusForbidden)
 		fmt.Fprintf(w, "Access Denied! Not allowed to access the resource\n")

--- a/pkg/git/server/server.go
+++ b/pkg/git/server/server.go
@@ -109,11 +109,11 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 
 	// The authentication was already validated, should not return error
-	_, jwtTokenString, _ := r.BasicAuth()
+	// _, jwtTokenString, _ := r.BasicAuth()
 	env := gitEnv{
-		Namespace:      namespace,
-		DeployName:     dp.Name,
-		UserJwtToken:   jwtTokenString,
+		Namespace:  namespace,
+		DeployName: dp.Name,
+		// UserJwtToken:   jwtTokenString,
 		GitHome:        h.cnf.GitHome,
 		KubeSvcHost:    h.cnf.Host,
 		GitAPIHostname: h.cnf.GitAPIHostname,
@@ -242,7 +242,7 @@ func gitCommand(env gitEnv, name string, args ...string) (*exec.Cmd, io.Reader) 
 	cmd.Env = append(os.Environ(),
 		fmt.Sprintf("NAMESPACE=%s", env.Namespace),
 		fmt.Sprintf("DEPLOY_NAME=%s", env.DeployName),
-		fmt.Sprintf("USER_JWT_TOKEN=%s", env.UserJwtToken),
+		// fmt.Sprintf("USER_JWT_TOKEN=%s", env.UserJwtToken),
 		fmt.Sprintf("KUBERNETES_SERVICE_HOST=%s", env.KubeSvcHost),
 		fmt.Sprintf("GIT_HOME=%s", env.GitHome),
 		fmt.Sprintf("GIT_API_HOSTNAME=%s", env.GitAPIHostname),

--- a/pkg/util/util.go
+++ b/pkg/util/util.go
@@ -1,19 +1,23 @@
 package util
 
 import (
+	"bytes"
 	"encoding/json"
 	"fmt"
 	"hash/adler32"
 	"io"
 	"net/http"
 	"strconv"
+	"time"
 
 	"io/ioutil"
 
+	jwt "github.com/dgrijalva/jwt-go"
 	"github.com/golang/glog"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	platform "kolihub.io/koli/pkg/apis/v1alpha1"
 )
 
 // Decoder known how to decode a io.Reader into a Kubernetes resource
@@ -36,6 +40,24 @@ func (d *Decoder) Decode(v runtime.Object) error {
 	}
 	// runtime
 	return runtime.DecodeInto(d.dec, data, v)
+}
+
+// GenerateNewJwtToken creates a new user token to allow machine-to-machine interaction
+func GenerateNewJwtToken(key, customer, org string, tokenType platform.TokenType, exp time.Time) (string, error) {
+	token := jwt.New(jwt.SigningMethodHS256)
+	claims := make(jwt.MapClaims)
+	// Set some claims
+	claims["customer"] = customer
+	claims["org"] = org
+	claims["kolihub.io/type"] = tokenType
+
+	// always convert to UTC time
+	claims["exp"] = exp.UTC().Unix() // claims["exp"] = time.Now().UTC().Add(time.Minute * 20).Unix()
+	claims["iat"] = time.Now().UTC().Unix()
+	token.Claims = claims
+
+	// Sign and get the complete encoded token as a string
+	return token.SignedString(bytes.NewBufferString(key).Bytes())
 }
 
 // StrategicMergePatch creates a strategic merge patch and merge with the original object

--- a/rootfs/Dockerfile.controller
+++ b/rootfs/Dockerfile.controller
@@ -1,4 +1,4 @@
-FROM quay.io/koli/ubuntu:16.04
+FROM quay.io/koli/base-os
 
 RUN adduser --system \
     --shell /bin/bash \

